### ISSUE TITLE
Recieve and Delete large amount of messages does not work

### ISSUE THREAD

**Issue by deadlydog** [paolosalvatori/ServiceBusExplorer#60]
If you choose to Receive and Delete 50,000 messages from a queue/subscriptions/deadletter it will typically only 2000 - 4000 messages, instead of the specified amount. The amount of messages that gets removed is different each time; sometimes it's 2900 messages, the next time it might be 3600. Doing Receive and Delete All has the same problem as well. I'm not sure if the tool is timing out or how it determines when to stop receiving messages, but it seems it does hit some limit. This makes it tough to clear out our deadletter queue when it has a 200,000 messages in it. We've written a console app to do this for us, but it would be great if it just worked inside of Service Bus Explorer.

**Comment by paolosalvatori** [comment]
Can you see any exception in the log? An alternative way to consume messages from a queue or subscription is using a Listener. I'm pretty sure that this consumes all of the messages, no matter what. Please right click the entity and start a listener, then let me know. Thanks!

**Comment by SeanFeldman** [comment]
@deadlydog did that work for you?

**Comment by deadlydog** [comment]
Hey guys, sorry for not getting back to you sooner, but I've been busy with other things.

Here is what is shown in the Log after doing a "Receive and Delete" "All" messages in the deadletter queue (that has 400,000+ messages in it) of one of my topic subscriptions:

```
<10:54:26> [64510] messages received from the deadletter queue of the subscription [RqPlatformPullService.PH].
```

It does not show any exceptions or errors, just that one line.

I didn't realize that creating a listener was an option/feature. I tried using it, but I couldn't figure out how to hook it up to the deadletter queue, only the regular subscription, so it seems I can't use it for this particular use case.

Any other suggestions? Thanks!

**Comment by SeanFeldman** [comment]
@deadlydog listener doesn't work for DLQs. Not sure if that was intentional or not.

**Comment by paolosalvatori** [comment]
Hi @deadlydog that's a scenario I never considered, I mean starting a listener on a DLQ rather than a regular queue or subscription. Time permitting, I'll implement this feature. Hopefully, more on this soon.

**Comment by deadlydog** [comment]
Thanks @paolosalvatori. While having a listener on a DLQ is an interesting feature request, it is a separate issue from the bug that I have reported. The bug is if I have 100,000 messages and want to receive (and maybe delete) the first 50,000, I can't do that. It always stops after the first few thousand for some reason. I can reproduce this problem every time, so I imagine you should be able to quite easily as well.

**Comment by paolosalvatori** [comment]
The problem could be a UI problem as I load messages into a DataViewGrid. Now, loading 100,000 messages into a UI control doesn't make a lot of sense as it's very hard to navigate through the messages. I will add the ability to disable loading messages into the grid and see if the problem goes away. In addition, I will check the limits of the DataGridView control

**Comment by paolosalvatori** [comment]
I implemented Purge Messages and Purge Deadletter Queue Messages commands for queues and subscriptions.

**Comment by deadlydog** [comment]
That's awesome that you added that new functionality @paolosalvatori, that will be super useful! However, it doesn't address the original bug which is when you want to only delete 50,000 or the 100,000 messages, there's no easy way to do it because the tool stops after deleting the first few thousand messages. I think this should be reopened, unless that issue was actually fixed as well and just not mentioned in this ticket. Thanks!

**Comment by deadlydog** [comment]
Hey @paolosalvatori I finally got around to downloading the new version of Service Bus Explorer. I tried out the new Purge Deadletter Queue Messages feature, and unfortunately it suffers from the same problem as the original bug I mentioned. If there are too many messages in the deadletter queue, it does not delete them all. 

I have a deadletter queue with 21,000 messages in it. The first time I clicked on the Purge Deadletter Queue Messages I got the spinner for a few seconds, but it didn't remove any messages. The 2nd time I tried I got the spinner for a few seconds and it did delete about 2,000 messages, but not all of them.

I'm using the latest Release, v4.0.87. This issue should be reopened.

**Comment by deadlydog** [comment]
@paolosalvatori or @SeanFeldman, could you reopen this issue please (even if you don't have time to get to it anytime soon), as the issue is still present. It's better to know that it's still an outstanding issue than to think it's been fixed when it hasn't. Maybe by having it on the `Open` issues list, somebody else will submit a PR to fix it. Thanks :)

**Comment by SeanFeldman** [comment]
@deadlydog how many messages did you have and on what type of entity when it failed?

**Comment by deadlydog** [comment]
@SeanFeldman The same behavior happens for Queues, Subscriptions, and DLQs. If you have 20,000+ messages on either type and try to clear them (either using the old `Receive And Delete` or the new `Purge` functions), you'll see the error. It's very easy to reproduce, and I've had other coworkers confirm the same behavior.

**Comment by deadlydog** [comment]
@SeanFeldman @paolosalvatori If you guys need more information to get this issue reopened please let me know. As I mentioned, we can easily reproduce this problem every time, so if you guys aren't seeing the problem I would be interested in finding what the difference is. We ended up creating our own console application specifically for clearing messages from a queue/subscription/DLQ (mostly used in our Testing environment), but it would be much more convenient to just use Service Bus Explorer for it since we are already using it for many other things. Thanks.

**Comment by SeanFeldman** [comment]
@deadlydog my bad, accidentally closed it as I was posting my comment to you 🙂

**Comment by paolosalvatori** [comment]
I'll look into this when I have time. @deadlydog do you see any error when trying to purge messages from a queue or subscription or at the end of the task you simply see that there are still some messages in the queue/subscription in question? Can you please tell me of the queue(s) or topic(s) in question is/are partitioned? Are they located in a standard or premium Service Bus namespace? Thanks!

**Comment by deadlydog** [comment]
Hey @paolosalvatori sorry it's taken me so long to respond to you on this one. I recorded a gif of me using the Purge DLQ functionality that will hopefully answer some of your questions. In this example I'm purging the DLQ of a topic subscription, but I see the same behaviour on queues as well. The queue's/subscriptions are not partitioned. I'm not certain if our service bus namespace is standard or premium, sorry.

![purging dlq not working](https://user-images.githubusercontent.com/1187140/31910308-53ec6ae6-b7fa-11e7-9960-7103140d4a6b.gif)

**Comment by deadlydog** [comment]
I will note though that I just tried Right-click Subscription -> Receive Deadletter Queue Messages -> select `Receive and Delete` and `All`, and that method actually did go through and delete all ~200,000 deadletter messages. That never worked the last time I tried it, but I may have been on an older version of the app last time I tried.

**Comment by paolosalvatori** [comment]
Thanks for the feedback @deadlydog maybe we should change the code to retry at least a couple of times before considering the operation finished. I'll review the code as soon as I've some spare time.

**Comment by ErikMogensen** [comment]
I have started to work on this. My plan is to solve it by:

- Gradually increasing the wait interval when receiving if no messages are received.

- Depending upon the number of messages in the queue/subscription/DLQ create several receive tasks to perform the purging quicker,

- Make up to three retries when receiving zero messages, but quit when the number of messages that existed in the entity when starting have been purged.

**Comment by PsMaster** [comment]
Topic subscriptions, if I try to use Purge DLQ it just doesn't delete anything...
[0] messages have been purged from the deadletter queue of the [sometopic/Subscriptions/2222_xxxx_654654] subscription in [2400] milliseconds.
Using Service Bus Explorer version 4.0.105

**Comment by SeanFeldman** [comment]
Would be so nice to have https://github.com/Azure/azure-service-bus/issues/1 implemented.
Perhaps all affected by this should up-vote it? 😉

**Comment by paolosalvatori** [comment]
Agreed. The tool just deletes messages from a topic, subscription or deadletter queue of a messaging entity. The code to delete messages has been created and tested years ago. @PsMaster, could you please repro the problem with another queue, possibly on a different namespace and confirm the issue? Thanks!

**Comment by PsMaster** [comment]
@paolosalvatori attaching a screenshot to show what's happening, if I try to purge any of the deadletters on the subscriptions it just spins around for few seconds and says 0 messages deleted.
![servicebus](https://user-images.githubusercontent.com/5575641/35094350-eb3a5094-fc3c-11e7-9712-72c11e0f2f91.PNG)

**Comment by SeanFeldman** [comment]
How many dead-lettered messages o you have in those subscriptions?

**Comment by PsMaster** [comment]
@SeanFeldman should it matter how many messages? Let's just say a lot, it's a big system and due so a bug with SignalR some messages are not picked up which results in deadletters so I have to manually go through and clear the messages from time to time so that it doesn't run out of space until we roll out a fix, which currently only works by deleting the whole subscription and is not ideal... We're talking about 1M+ over all the subscriptions, but up to ~200K per single subscription. What I used to notice before is the bug that when you do purge messages on normal messages it just clears out a random number of those, but for deadletters this doesn't work at all.

**Comment by radsolution** [comment]
Hello, I am getting a similar rproblem - I currently have about 950 deadlettered messages, but attempting to peek 100 returns only 207 with no reason why...
`<11:31:00> [207] messages peeked from the deadletter queue of the subscription [myQueue]`.

**Comment by ErikMogensen** [comment]
@radsolution, cool user name BTW, please let us know what values you have for the three settings that I put the red rectangle around in the figure below:
![image](https://user-images.githubusercontent.com/9644248/71358303-d060b480-2588-11ea-9d49-036ca3d67c92.png)
Also, is this a partitioned queue? Is it Basic, Standard or Premium?

Are you using the latest version? It was released in May this year. Do not look at the version number as it is dynamic in older Service Bus Explorer versions.

Increase Receive and Server timeout to 180 and see if that helps.

**Comment by radsolution** [comment]
Hello,
Thank you for your reply.
Here are my settings (your image was cut off above the red markings):

I think the queue (topic) is basic, not partitioned. The version number shown is 3.0.4
Thanks, Dave Radcliffe
    On Monday, 23 December 2019, 12:37:47 GMT, Erik Mogensen <notifications@github.com> wrote:  
 
 
@radsolution, cool user name BTW, please let us know what values you have for the three settings that I put the red rectangle around in the figure below:

Also, is this a partitioned queue? Is it Basic, Standard or Premium?

Are you using the latest version? It was released in May this year. Do not look at the version number as it is dynamic in older Service Bus Explorer versions.

Increase Receive and Server timeout to 180 and see if that helps.

**Comment by ErikMogensen** [comment]
Then @radsolution , you have an old version. Please upgrade to the latest from https://github.com/paolosalvatori/ServiceBusExplorer/releases and try again.

**Comment by radsolution** [comment]
Thank you.
    On Monday, 23 December 2019, 17:58:13 GMT, Erik Mogensen <notifications@github.com> wrote:  
 
 
Then @radsolution , you have an old version. Please upgrade to the latest from https://github.com/paolosalvatori/ServiceBusExplorer/releases and try again.

**Comment by ErikMogensen** [comment]
Assuming that the issue was resolved by using the latest version and therefore closing this.

**Comment by radsolution** [comment]
Yes, thank you.
    On Thursday, 16 January 2020, 15:46:51 GMT, Erik Mogensen <notifications@github.com> wrote:  
 
 
Closed #60.
